src/shims/posix/fs.rs

```diff
@@ -412,24 +447,29 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         let fd = this.read_scalar(fd_op)?.to_i32()?;
 
-        if let Some(FileHandle { file, writable }) = this.machine.file_handler.handles.remove(&fd) {
-            // We sync the file if it was opened in a mode different than read-only.
-            if writable {
-                // `File::sync_all` does the checks that are done when closing a file. We do this to
-                // to handle possible errors correctly.
-                let result = this.try_unwrap_io_result(file.sync_all().map(|_| 0i32));
-                // Now we actually close the file.
-                drop(file);
-                // And return the result.
-                result
-            } else {
-                // We drop the file, this closes it but ignores any errors produced when closing
-                // it. This is done because `File::sync_all` cannot be done over files like
-                // `/dev/urandom` which are read-only. Check
-                // https://github.com/rust-lang/miri/issues/999#issuecomment-568920439 for a deeper
-                // discussion.
-                drop(file);
-                Ok(0)
+        if let Some(file_descriptor) = this.machine.file_handler.handles.remove(&fd) {
```

Comment:
Why does `as_file_handle` return a `Result`? If it returned an `Option, you could
```suggestion
        if let Some(file_descriptor) = this.machine.file_handler.handles.remove(&fd).and_then(FileDescriptor::as_file_handle) {
```

and save yourself the additional `this.handle_not_found()`

Reply:
resolving, because it just makes things more complex